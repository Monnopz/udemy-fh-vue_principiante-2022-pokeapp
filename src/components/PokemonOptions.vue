<template>
  <div>
    <div class="options-container">
        <ul>
            <!-- El equipo de Vue agrupo variables, ojetos y metodos con simbolo de dolar $; en este caso $emit -->
            <!-- $emit(nombreMetodo, valoraenviar) -->
            <!-- Como buena practica de parte del equipo de Vue, al emitir usar upperCamelCase (selectionPokemon) y al escuchar el evento usar keba-cas(selection-pokemon) -->
            <!-- Para escuchar un evento simplemente es con arroba (@); @selection-pokemon="metodoDelPadreAEjecutar" -->
            <li v-for="pokemon in pokemons" :key="pokemon.id" @click="$emit('selectionPokemon', pokemon.id)">{{pokemon.name}}</li>
        </ul>
    </div>
  </div>
</template>

<script>
export default {
    name: 'PokemonOptions',
    props: {
        pokemons: {
            type: Array,
            required: true
        }
    }
}
</script>

<style scoped>
/* Pokemon Options */
ul {
    list-style-type: none;
    padding-left: 0;
}
li {
    background-color: white;
    border-radius: 5px;
    border: 1px solid rgba(0, 0, 0, 0.2);
    cursor: pointer;
    margin-bottom: 10px;
    width: 250px;
}

li:hover {
    background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
    display: flex;
    justify-content: center;
}
</style>