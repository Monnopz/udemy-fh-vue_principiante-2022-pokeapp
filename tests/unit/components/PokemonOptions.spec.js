import { shallowMount } from '@vue/test-utils'
import PokemonOptions from '@/components/PokemonOptions'

//Este import es un array creado en un fichero para tomarlo en este archivo y utilizarlo
import { mockPokemons } from '../mocks/pokemons.mock'

describe('PokemonOptions Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonOptions, {
            props: {
                pokemons: mockPokemons
            }
        })
    })

    test('debe de hacer match con el snapshot', () => {

        // expect(wrapper.html()).toMatchInlineSnapshot() ; pone dentro de los parentesis aqui mismo un html inline con el que se compara
        expect(wrapper.html()).toMatchSnapshot()

    })

    test('debe de mostrar las 4 opciones correctamente', () => {

        const liTags = wrapper.findAll('li')
        expect(liTags.length).toBe(4)

        //Para sacar el texto de un elemento HTML
        expect(liTags[0].text()).toBe('bulbasaur')
        expect(liTags[1].text()).toBe('ivysaur')
        expect(liTags[2].text()).toBe('venusaur')
        expect(liTags[3].text()).toBe('charmander')

    });

    test('debe de emitir "selection" con sus respectivos parametros al hacer click', () => {
        
        const [li1, li2, li3, li4] = wrapper.findAll('li')

        li1.trigger('click') //Estimulo; da clic al boton
        li2.trigger('click') //Estimulo; da clic al boton
        li3.trigger('click') //Estimulo; da clic al boton
        li4.trigger('click') //Estimulo; da clic al boton

        expect(wrapper.emitted('selectionPokemon').length).toBe(4) //verifca que se haya emitido el evento selectionPokemon (el length por que quiere decir que si se emitio, tendra un valor en array)
        expect(wrapper.emitted('selectionPokemon')[0][0]).toStrictEqual(1)
        // expect(wrapper.emitted('selectionPokemon')[0]).toStrictEqual([1])

        expect(wrapper.emitted('selectionPokemon').length).toBe(4) //verifca que se haya emitido el evento selectionPokemon (el length por que quiere decir que si se emitio, tendra un valor en array)
        expect(wrapper.emitted('selectionPokemon')[1][0]).toStrictEqual(2)
        // expect(wrapper.emitted('selectionPokemon')[0]).toStrictEqual([2])

        expect(wrapper.emitted('selectionPokemon').length).toBe(4) //verifca que se haya emitido el evento selectionPokemon (el length por que quiere decir que si se emitio, tendra un valor en array)
        expect(wrapper.emitted('selectionPokemon')[2][0]).toStrictEqual(3)
        // expect(wrapper.emitted('selectionPokemon')[0]).toStrictEqual([3])

        expect(wrapper.emitted('selectionPokemon').length).toBe(4) //verifca que se haya emitido el evento selectionPokemon (el length por que quiere decir que si se emitio, tendra un valor en array)
        expect(wrapper.emitted('selectionPokemon')[3][0]).toStrictEqual(4)
        // expect(wrapper.emitted('selectionPokemon')[0]).toStrictEqual([4])

    });

})