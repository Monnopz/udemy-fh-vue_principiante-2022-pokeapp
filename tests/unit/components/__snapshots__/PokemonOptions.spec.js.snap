// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions Component debe de hacer match con el snapshot 1`] = `
<div>
  <div class="options-container">
    <ul>
      <!-- El equipo de Vue agrupo variables, ojetos y metodos con simbolo de dolar $; en este caso $emit -->
      <!-- $emit(nombreMetodo, valoraenviar) -->
      <!-- Como buena practica de parte del equipo de Vue, al emitir usar upperCamelCase (selectionPokemon) y al escuchar el evento usar keba-cas(selection-pokemon) -->
      <!-- Para escuchar un evento simplemente es con arroba (@); @selection-pokemon="metodoDelPadreAEjecutar" -->
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div>
</div>
`;
